File: UltimateTicTacToe/tic_tac_toe_gui.py
from gui import GuiObject, animations
from math import tau

class TicTacToeGui(GuiObject):

    @staticmethod
    def draw_nought(gfx, x, y, width, height, color, anim_value=1, thickness=3):
        gfx.arc(x + width / 2, y + height / 2, width / 2, 0, int(360 * anim_value), line_color=color, line_thickness=thickness)

    @staticmethod
    def draw_cross(gfx, x, y, width, height, color, anim_value=1, thickness=3):
        gfx.line((x, y), (x + width * anim_value, y + height * anim_value), color, thickness)
        gfx.line((x + width, y), (x + width * (1 - anim_value), y + height * anim_value), color, thickness)

    def __init__(self, x, y, width, height, tictactoe):

        GuiObject.__init__(self, x, y, width, height)

        self.tictactoe = tictactoe
        self.win_anim = animations.Animation(500)
        self.animations = []
        for row in range(self.tictactoe.rows):
            self.animations.append([])
            for col in range(self.tictactoe.cols):
                self.animations[row].append(animations.Animation(500))

    def mouse_down(self, x, y):
        if self.tictactoe.game_over():
            enabled = False
        else:
            row = int(y // (self.height / self.tictactoe.rows))
            col = int(x // (self.width / self.tictactoe.cols))

            if self.tictactoe.play(row, col):
                self.animations[row][col].start()
            if self.tictactoe.winner:
                self.win_anim.start()

    def draw(self, gfx):
        # background
        gfx.background((255, 255, 255))

        # drawing seperation line
        if self.tictactoe.winner == self.tictactoe.NOUGHT:
            color = (255, 0, 0)
        elif self.tictactoe.winner == self.tictactoe.CROSS:
            color = (0, 0, 255)
        elif self.tictactoe.winner == self.tictactoe.EMPTY:
            color = (0, 255, 0)
        else:
            color = (0, 0, 0)

        hori = self.height / self.tictactoe.rows
        for row in range(1, self.tictactoe.rows):
            gfx.line((0, hori * row), (self.width, hori * row), color, 3)
        verti = self.width / self.tictactoe.cols
        for col in range(1, self.tictactoe.cols):
            gfx.line((verti * col, 0), (verti * col, self.height), color, 3)

        # drawing noughts and crosses
        space = .05
        width = self.width / self.tictactoe.cols
        height = self.height / self.tictactoe.rows

        for row in range(self.tictactoe.rows):
            for col in range(self.tictactoe.cols):
                if self.tictactoe.board[row][col] != self.tictactoe.EMPTY:
                    x = self.width / self.tictactoe.cols * col
                    x += width * space
                    y = self.height / self.tictactoe.rows * row
                    y += height * space

                    w = width * (1 - space * 2)
                    h = height * (1 - space * 2)
                    anim_value = self.animations[row][col].value()
                    if self.tictactoe.board[row][col] == self.tictactoe.NOUGHT:
                        self.draw_nought(gfx, x, y, w, h, (255, 0, 0), anim_value)
                    else:
                        self.draw_cross(gfx, x, y, w, h, (0, 0, 255), anim_value)

        # drawing win line
        if self.win_anim.started():
            if self.tictactoe.winner == self.tictactoe.CROSS:
                self.draw_cross(gfx, self.x, self.y, self.height, self.width, (0, 0, 255), self.win_anim.value(), 100)
            else:
                self.draw_nought(gfx, self.x, self.y, self.height, self.width, (255, 0, 0), self.win_anim.value(), 100)

File: UltimateTicTacToeCS/TicTacToeGui.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace UltimateTicTacToeCS
{
    public partial class TicTacToeGui : GuiControl
    {
        public static void DrawCross(Graphics gfx, RectangleF rect, Color color, int width, float animValue = 1)
        {
            var location = rect.Location;
            var size = rect.Size;

            if (animValue < .5f)
            {
                gfx.DrawLine(new Pen(new SolidBrush(color), width), location,
                    new PointF(location.X + size.Width * animValue * 2, location.Y + size.Height * animValue * 2));
            }
            else
            {
                gfx.DrawLine(new Pen(new SolidBrush(color), width), location,
                    new PointF(location.X + size.Width, location.Y + size.Height));

                animValue -= .5f;
                gfx.DrawLine(new Pen(new SolidBrush(color), width), new PointF(location.X + size.Width, location.Y),
                    new PointF(location.X + size.Width * (1 - animValue * 2), location.Y + size.Height * animValue * 2));
            }
        }

        public static void DrawNought(Graphics gfx, RectangleF rect, Color color, int width, float animValue = 1)
        {
            gfx.DrawArc(new Pen(new SolidBrush(color), width), rect, -90, 360 * animValue);
        }

        public static void DrawAvailable(Graphics gfx, RectangleF rect, Color color, float animValue = 1, float mouseAnim = 0)
        {
            mouseAnim *= .25f;
            rect = new RectangleF(rect.X + rect.Width / 2 * (1 - animValue - mouseAnim), rect.Y + rect.Height / 2 * (1 - animValue - mouseAnim),
                rect.Width * (animValue + mouseAnim), rect.Height * (animValue + mouseAnim));
            gfx.FillEllipse(new SolidBrush(color), rect);
        }

        public TicTacToe TicTacToe { get; private set; }
        public bool MouseClickEnabled { get; set; }
        public bool ShowLastMove { get; set; }

        private int WinLineWidth => 50 * Math.Min(Width, Height) / 300;
        private Animation[,] animations;
        private Animation winAnimation;
        private Animation enabled;
        private Animation mouseOn;
        private Point sqrMouse;

        public TicTacToeGui()
        {
            InitializeComponent();

            MouseClickEnabled = true;
            ShowLastMove = true;
            enabled = new Animation(250);
            NewGame();
        }
        public TicTacToeGui(TicTacToe board) : this()
        {
            TicTacToe = board;
        }

        public void NewGame()
        {
            animations = new Animation[TicTacToe.ROWS, TicTacToe.COLS];
            TicTacToe = new TicTacToe();
            sqrMouse = new Point(-1, -1);

            for (int row = 0; row < TicTacToe.ROWS; ++row)
            {
                for (int col = 0; col < TicTacToe.COLS; ++col)
                {
                    animations[row, col] = new Animation(500);
                }
            }

            winAnimation = new Animation(500);
            mouseOn = new Animation(250);
            enabled.Start();
        }

        protected override void DrawSquare(int row, int col, RectangleF rect)
        {
            // Draw last move.
            if (ShowLastMove && TicTacToe.LastMove[0] == row && TicTacToe.LastMove[1] == col)
            {
                /*float x = ColWidth * col;
                float y = RowHeight * row;*/
                gfx.FillRectangle(new SolidBrush(Options.Theme.LastMove), rect);
            }

            if (TicTacToe.Board[row, col] == TicTacToe.SqrState.Cross)
            {
                DrawCross(gfx, rect, Options.Theme.Cross, LineWidth, animations[row, col].Value);
            }
            else if (TicTacToe.Board[row, col] == TicTacToe.SqrState.Nought)
            {
                DrawNought(gfx, rect, Options.Theme.Nought, LineWidth, animations[row, col].Value);
            } 
            else if (Enabled && !TicTacToe.GameOver)
            {
                rect = new RectangleF(rect.X + rect.Width / 4, rect.Y + rect.Height / 4, rect.Width / 2, rect.Height / 2);

                float mouseAnim = 0;
                if (sqrMouse.X == row && sqrMouse.Y == col)
                {
                    mouseAnim += mouseOn.Value;
                }
                
                DrawAvailable(gfx, rect, Options.Theme.AvailableSquare, enabled.Value, mouseAnim);
            }
        }

        public override Image Draw()
        {
            var bm = base.Draw();

            // Draw win animation.
            if (TicTacToe.Winner == TicTacToe.WinState.Cross)
            {
                DrawCross(gfx, new RectangleF(0, 0, Width, Height), Options.Theme.Cross, WinLineWidth, winAnimation.Value);
            }
            else if (TicTacToe.Winner == TicTacToe.WinState.Nought)
            {
                DrawNought(gfx, new RectangleF(WinLineWidth / 2, WinLineWidth / 2, Width - WinLineWidth, Height - WinLineWidth), Options.Theme.Nought, WinLineWidth, winAnimation.Value);
            }

            return bm;
        }

        public void Played(int row, int col)
        {
            animations[row, col].Start();

            if (TicTacToe.GameOver)
            {
                winAnimation.Start();
            }

            sqrMouse.X = -1;
        }

        private void MouseClicked(object sender, MouseEventArgs e)
        {
            if (MouseClickEnabled)
            {
                if (e.Button == MouseButtons.Left)
                {
                    if (TicTacToe.GameOver)
                    {
                        NewGame();
                    }
                    else
                    {
                        int row = e.Y / (Height / TicTacToe.ROWS);
                        int col = e.X / (Width / TicTacToe.COLS);

                        if (TicTacToe.Play(row, col))
                        {
                            Played(row, col);
                        }
                    }
                }
                else
                {
                    if (!Animation.Paused)
                    {
                        Animation.Pause();
                    }
                    else
                    {
                        Animation.Resume();
                    }
                }
            }
        }

        private void MouseMoved(object sender, MouseEventArgs e)
        {
            int row = e.Y / (Height / TicTacToe.ROWS);
            int col = e.X / (Width / TicTacToe.COLS);

            if (!TicTacToe.GameOver && TicTacToe.InBounds(row, col) && TicTacToe.Board[row, col] == TicTacToe.SqrState.Empty)
            {
                if (sqrMouse.X != row || sqrMouse.Y != col)
                {
                    sqrMouse = new Point(row, col);
                    mouseOn.Start();
                }
            }
            else
            {
                sqrMouse.X = -1;
            }
        }

        private void MouseLeaved(object sender, EventArgs e)
        {
            sqrMouse = new Point(-1, -1);
        }

        private void TicTacToeGuiEnabledChanged(object sender, EventArgs e)
        {
            if (Enabled)
            {
                enabled.Start();
            }
        }
    }
}
